File: Java/ShippableProduct.java
public abstract class ShippableProduct extends Product implements Shippable {
    protected double weight;
    
    public ShippableProduct(String name, double price, int quantity, double weight) {
        super(name, price, quantity);
        this.weight = weight;
    }
    
    @Override
    public double getWeight() {
        return weight;
    }
    
    @Override
    public boolean requiresShipping() {
        return true;
    }
}

File: javascript/ShippableProduct.js
// javascript/ShippableProduct.js
import { Product } from './Product.js';
import { IShippable } from './Shippable.js';

export class ShippableProduct extends Product {
    constructor(name, price, quantity, weight) {
        super(name, price, quantity);
        if (this.constructor === ShippableProduct) {
            throw new Error("Abstract classes can't be instantiated.");
        }
        this.weight = weight;
    }

    getWeight() {
        return this.weight;
    }

    requiresShipping() {
        return true;
    }
}
